cpp/CMakeLists.txt

```diff
@@ -100,96 +106,129 @@ endif()
 # add third party dependencies using CPM
 rapids_cpm_init()
 
-# thrust and libcudacxx need to be before cuco!
-include(cmake/thirdparty/get_thrust.cmake)
 include(cmake/thirdparty/get_rmm.cmake)
 include(cmake/thirdparty/get_libcudacxx.cmake)
 include(cmake/thirdparty/get_cuco.cmake)
+# CPMFindPackage(faiss) so it's available for raft consumers to use
+include(cmake/thirdparty/get_faiss.cmake)
```

Comment:
I'm a little confused (after being pinged about this by @divyegala ), but wouldn't moving this out of the conditional, make it so that every consumer of raft calls cpm find faiss? how does cmake know to not do the get_faiss.cmake logic when not linking to `nn` or `distance` components?

Reply:
For context, we saw this happen in a consumer library which was neither linking to `nn` or `distance` but started building FAISS by default

Reply:
You are correct when building from source currently `get_faiss` will be called.

CMake doesn't have any mechanisim to allow lazy/deferred find/cpm calls based on a targets usage. So we can't delay looking for `faiss` based on the usage of `raft::nn`.

Since `raft::nn` can be used without any compilation, and requires faiss we are in an ambigous situation.

What I have done is add a `RAFT_ENABLE_NN_DEPENDENCIES` option whose default value is the same as `RAFT_COMPILE_LIBRARIES`. CPM consumers of raft ( cuml/cugraph ) which don't want to use the `raft::nn` can set `RAFT_ENABLE_NN_DEPENDENCIES` to OFF in the CPM options to disable faiss searches.

Reply:
Can confirm this option works well! Thanks

Reply:
@robertmaynard, you had mentioned (offline) that we should be able to do something like `find_package(raft COMPONENTS nn)` in order to bring in the FAISS dependency (cuml would use), while `find_package(raft)` won't search for FAISS at all (which cugraph would use). Is this still something we are able to do in order to simplify the dependency process for consuming libraries?

Reply:
Perfect!